File: packages/python-sdk/seghro/langchain_callback.py
"""LangChain callback handler for Seghro observability."""

from __future__ import annotations
import time
from seghro.client import SeghroClient

class SeghroCallbackHandler:
    def __init__(self, api_key: str, agent_name: str, endpoint: str = None, debug: bool = False):
        self.client = SeghroClient(api_key=api_key, agent_name=agent_name, agent_framework="LangChain", endpoint=endpoint or "https://seghro.dev/api/otlp/v1/traces", debug=debug)
        self.spans = []
        self.start_time = 0

    def on_chain_start(self, serialized, inputs, **kwargs):
        self.start_time = time.time()
        self.spans = []

    def on_llm_end(self, response, **kwargs):
        pass

    def on_chain_end(self, outputs, **kwargs):
        duration = (time.time() - self.start_time) * 1000
        self.client.ingest_trace(status="success", duration=duration, spans=self.spans)

    def on_chain_error(self, error, **kwargs):
        duration = (time.time() - self.start_time) * 1000
        self.client.ingest_trace(status="error", duration=duration, spans=self.spans)

File: packages/js-sdk/src/langchain-callback.ts
import { SeghroClient } from './seghro-client';

export class SeghroCallbackHandler {
  private client: SeghroClient;
  private spans: any[] = [];
  private startTime = 0;

  constructor(config: { apiKey: string; agentName: string; endpoint?: string; debug?: boolean }) {
    this.client = new SeghroClient({ ...config, agentFramework: 'LangChain' });
  }

  async handleChainStart() { this.startTime = Date.now(); this.spans = []; }
  async handleLLMStart(llm: any, prompts: string[]) { this.spans.push({ name: 'llm_call', type: 'model', status: 'success', duration: 0, model: llm.id?.at(-1) || 'unknown' }); }
  async handleLLMEnd(output: any) {
    const last = this.spans[this.spans.length - 1];
    if (last?.name === 'llm_call') { last.duration = Date.now() - this.startTime; last.inputTokens = output.llmOutput?.tokenUsage?.promptTokens || 0; last.outputTokens = output.llmOutput?.tokenUsage?.completionTokens || 0; }
  }
  async handleChainEnd() { await this.client.ingestTrace({ status: 'success', duration: Date.now() - this.startTime, inputTokens: this.spans.reduce((s, sp) => s + (sp.inputTokens || 0), 0), outputTokens: this.spans.reduce((s, sp) => s + (sp.outputTokens || 0), 0), spans: this.spans }); }
  async handleChainError(error: Error) { await this.client.ingestTrace({ status: 'error', duration: Date.now() - this.startTime, spans: this.spans, metadata: { error: error.message } }); }
}
